upstream/gitlab.go

```diff
@@ -0,0 +1,134 @@
+package upstream
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/go-errors/errors"
+)
+
+// Self-hosted Gitlab instances use different domain names
+type gitLab struct {
+	domain     string
+	owner      string
+	repository string
+}
+
+func (g gitLab) String() string {
+	return g.domain + "/" + g.owner + "/" + g.repository
+}
+
+func (g gitLab) encoded() string {
+	// Gitlab requires the owner + repository to be url-encoded together
+	return url.PathEscape(g.owner) + "%2F" + url.PathEscape(g.repository)
+}
+
+func (g gitLab) releasesURL() string {
+	// API documentation: https://docs.gitlab.com/ee/api/tags.html#list-project-repository-tags
+	// Note that the second %s must be url-encoded (see gitLab.encoded())
+	return fmt.Sprintf("https://%s/api/v4/projects/%s/repository/tags", g.domain, g.encoded())
+}
+
+func (g gitLab) errorWrap(err error) error {
+	return errors.WrapPrefix(err, "Failed to obtain Gitlab tag for "+g.String()+" from "+g.releasesURL(), 0)
+}
+
+func (g gitLab) errorNotFound() error {
+	return errors.Errorf("No Gitlab release found for %s on %s", g, g.releasesURL())
+}
+
+// A Gitlab commit descriptor
+type gitLabCommit struct {
```

Comment:
Please remove all type structs, fields and comments that we are not evaluating in `latestVersion`:

```go
type gitLabTag struct {
	Name    string `json:"name"`
	Message string `json:"message"`
}

type gitLabMessage struct {
	Message string `json:"message"`
}
```